### PR TITLE
[KIE-790]  Build error on RuleUnit multiple OOPath lines

### PR DESCRIPTION
**Pull request by tkobayas**
https://github.com/apache/incubator-kie-issues/issues/790

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request checks</b>  
  Please add comment: <b>Jenkins retest this</b>

- for a <b>specific pull request check</b>  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] tests</b>

- for a <b>full downstream build</b> 
  - for <b>jenkins</b> job: please add comment: <b>Jenkins run fdb</b>
  - for <b>github actions</b> job: add the label `run_fdb`

- <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

- <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>

- for <b>quarkus branch checks</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins run quarkus-branch</b>

- for a <b>quarkus branch specific check</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] quarkus-branch</b>

- for <b>quarkus main checks</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins run quarkus-main</b>

- for a <b>specific quarkus main check</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] quarkus-main</b>

- for <b>quarkus lts checks</b>  
  Run checks against Quarkus lts branch  
  Please add comment: <b>Jenkins run quarkus-lts</b>

- for a <b>specific quarkus lts check</b>  
  Run checks against Quarkus lts branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] quarkus-lts</b>

- for <b>native checks</b>  
  Run native checks  
  Please add comment: <b>Jenkins run native</b>

- for a <b>specific native check</b>  
  Run native checks 
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] native</b>

 - for <b>native lts checks</b>  
  Run native checks against quarkus lts branch
  Please add comment: <b>Jenkins run native-lts</b>

- for a <b>specific native lts check</b>  
  Run native checks against quarkus lts branch
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|kogito-apps|kogito-examples] native-lts</b>
</details>

<details>
<summary>
How to backport a pull request to a different branch?
</summary>

In order to automatically create a **backporting pull request** please add one or more labels having the following format `backport-<branch-name>`, where `<branch-name>` is the name of the branch where the pull request must be backported to (e.g., `backport-7.67.x` to backport the original PR to the `7.67.x` branch).

> **NOTE**: **backporting** is an action aiming to move a change (usually a commit) from a branch (usually the main one) to another one, which is generally referring to a still maintained release branch. Keeping it simple: it is about to move a specific change or a set of them from one branch to another.

Once the original pull request is successfully merged, the automated action will create one backporting pull request per each label (with the previous format) that has been added.

If something goes wrong, the author will be notified and at this point a manual backporting is needed.

> **NOTE**: this automated backporting is triggered whenever a pull request on `main` branch is labeled or closed, but both conditions must be satisfied to get the new PR created.
</details>

### CODE REVIEW

File: drools-drl/drools-drl-parser/pom.xml [hunk 1]
```diff
@@ -127,47 +127,47 @@
                                             <path refid="maven.compile.classpath"/>
                                         </path -->
                                         <echo message="Generating Lexer (${antlr.debugParser})"/>
+<!--                                        <java classname="org.antlr.Tool" fork="true">-->
+<!--                                            <jvmarg line="-Xmx750M -Xincgc"/>-->
+<!--                                            <arg line="${antlr.debugParser} -lib src/main/resources/org/drools/compiler/lang src/main/resources/org/drools/compiler/lang/DRL5Lexer.g"/>-->
+<!--                                            <classpath refid="maven.compile.classpath"/>-->
+<!--                                        </java>-->
                                         <java classname="org.antlr.Tool" fork="true">
```
Comment (tkobayas) [inline drools-drl/drools-drl-parser/pom.xml:135]: I commented out unused Antlr gen tasks. If you want to use them, you would need to fix the file paths to keep up with past refactoring directory changes.

File: drools-drl/drools-drl-parser/pom.xml [hunk 2]
```diff
@@ -127,47 +127,47 @@
...
-                                            <jvmarg line="-Xmx750M -Xincgc"/>
-                                            <arg line="${antlr.debugParser} -lib src/main/resources/org/drools/compiler/lang src/main/resources/org/drools/compiler/lang/DRL5Lexer.g"/>
-                                            <classpath refid="maven.compile.classpath"/>
-                                        </java>
-                                        <java classname="org.antlr.Tool" fork="true">
-                                            <jvmarg line="-Xmx750M -Xincgc"/>
-                                            <arg line="${antlr.debugParser} -lib src/main/resources/org/drools/compiler/lang src/main/resources/org/drools/compiler/lang/DRL6Lexer.g"/>
+                                            <jvmarg line="-Xmx750M"/>
+                                            <arg line="${antlr.debugParser} -lib src/main/resources/org/drools/drl/parser src/main/resources/org/drools/drl/parser/DRL6Lexer.g"/>
```
Comment (tkobayas) [inline drools-drl/drools-drl-parser/pom.xml:137]: My change is basically on `DRL6Expressions.g`, but I needed to also enable `DRL6Lexer.g` gen to use `DRL6Lexer.tokens`

File: drools-drl/drools-drl-parser/src/main/resources/org/drools/drl/parser/DRL6Expressions.g [hunk 3]
```diff
@@ -567,7 +574,7 @@ xpathSeparator
     ;
 
 xpathPrimary returns [BaseDescr result]
-    : xpathChunk+
+    : xpathChunk ({notStartWithNewline()}? xpathChunk)*
```
Comment (tkobayas) [inline drools-drl/drools-drl-parser/src/main/resources/org/drools/drl/parser/DRL6Expressions.g:577]: This parser rule means "if a line starts with `/`, it is considered as a new OOPath, not a continuation of the previous OOPath."

**Comment by tkobayas** [comment]
DRL considers a new line is just `WS`, which is not used as a delimiter of the syntax.

So it was valid to parse
```
        /measurements[ $id: id == "color", $colorVal : val ]
	/sensors[ id == $id ]
```
as one constraint:
```
/measurements[ $id: id == "color", $colorVal : val ]/sensors[ id == $id ]
```

XPath spec accept a line break: https://www.w3.org/TR/2017/REC-xpath-31-20170321/#id-eol-handling
, but the spec explains about only one XPath expression.

If we want to consider that they are 2 OOPath patterns, we need to use a new line as a delimiter. So this fix is exceptional.

As a fundamental fix, IMO, we need to introduce an explicit syntax to begin an OOPath. e.g. `$/`
For example)
```
        $/measurements[ $id: id == "color", $colorVal : val ]
	$/sensors[ id == $id ]
```

WDYT? @mariofusco

**Comment by mariofusco** [comment]
@tkobayas This is a problem that I discussed in the past also with @lucamolteni and @tarilabs and we never found a solution. I think yours is very clever and fixes an actual unexpected problem for many users, so I'm in favor for bringing it in.

For what regards having a proper oopath separator, actually we already have the `;` which in my opinion is the most obvious one., so I wouldn't introduce any other. In essence if you write 
```
/measurements[ $id: id == "color", $colorVal : val ]; // <-- note the semicolon at the end of this line 
/sensors[ id == $id ]
```
it should already work. However I just searched our documentation to check if we're mentioning this and indeed I couldn't find any reference to it, so it would be nice to also add it to the docs.

**Comment by tkobayas** [comment]
Thank you, Mario! @mariofusco @lucamolteni Please review.

> For what regards having a proper oopath separator, actually we already have the `;`
> ...
> However I just searched our documentation to check if we're mentioning this and indeed I couldn't find any reference to it, so it would be nice to also add it to the docs.

Oh, thanks. It worked! I filed https://github.com/apache/incubator-kie-issues/issues/803 for documentation and will work on it :)

**Review by mariofusco — APPROVED** [review]

**Review by lucamolteni — APPROVED** [review]

**Pull request merged**